### PR TITLE
cisco_interface: interface lookup performance enhancement

### PR DESCRIPTION
**Pull request by chrisvanheuveln**
New PR based on original #557       _(prev PR was not picking up latest commits)_

#### Summary
- The existing behavior in `cisco_interface` provider code's `prefetch` method is to collect state for all interfaces on the device and to cache the data for other interface lookups. This approach makes sense in many cases but may result in unnecessarily long processing time for use cases where there are thousands of interfaces while the manifest is only interested in a small subset of interfaces. (Ref: https://github.com/cisco/cisco-network-puppet-module/issues/496)
- This PR creates a compromise by adding a per-interface lookup path to `prefetch` that is invoked when the manifest's `cisco_interface` resource count is below an auto-selected threshold value.
  - Testing has shown that the threshold sweet spot is roughly 15 - 20% of the total number of interfaces. This change uses 15% to err on the side of caution.
  - This percentage is based on the point where the total query times are nearly equal; in other words the amount of time required to manage a given number of manifest resources is the same for both querying for all interfaces versus querying for each managed interface separately.
  - Sample test results showing optimum threshold value / total device interfaces:
    - 20 / 100  (20.0%)
    - 30 / 200  (15.0%)
    - 100 / 600  (16.6%)
    - 200 / 1000 (20.0%)

- Supporting changes have already been made to the node_utilities interface code.
  - https://github.com/cisco/cisco-network-node-utils/pull/630
  - https://github.com/cisco/cisco-network-node-utils/pull/635
- Beaker test `test_interface_threshold.rb` was created to verify that the two code-paths are exercised when the managed count is on both sides of the threshold.
- There were a number of interface-related deprecated properties remaining in the `cisco_interface` provider. These have been removed.
- Added an rspec stub fix to `cisco_nexus_spec.rb` that was broken by #550

### CODE REVIEW

**Comment by mikewiebe** [comment]
General comment:  We should probably run a full regression test with these changes just to make sure there are no hidden issues.

File: lib/puppet/provider/cisco_interface/cisco.rb [hunk 1]
```diff
@@ -203,24 +186,41 @@ def self.properties_get(interface_name, nu_obj)
...
+    show_run_int_threshold = Cisco::Interface.interface_count * 0.15
+    if resources.keys.length > show_run_int_threshold
+      interfaces = instances
+      info "[prefetch all interfaces: (found: #{interfaces.length})]"
+      resources.keys.each do |name|
+        provider = interfaces.find { |intf| intf.instance_name == name }
+        resources[name].provider = provider unless provider.nil?
+      end
+    else
+      info "[prefetch each interface independently] threshold: #{show_run_int_threshold}"
```
[the remark this answers is not in the document]
Comment (chrisvanheuveln) [reply lib/puppet/provider/cisco_interface/cisco.rb]: disregarding based on offline discussion

File: lib/puppet/provider/cisco_interface/cisco.rb [hunk 2]
```diff
@@ -203,24 +186,41 @@ def self.properties_get(interface_name, nu_obj)
...
+    all_intf = single_intf ? false : true
     interfaces = []
-    Cisco::Interface.interfaces.each do |interface_name, nu_obj|
+    Cisco::Interface.interfaces(nil, single_intf).each do |interface_name, nu_obj|
       begin
         # Some interfaces cannot or should not be managed by this type.
         # - NVE Interfaces (managed by cisco_vxlan_vtep type)
         next if interface_name.match(/nve/i)
-        interfaces << properties_get(interface_name, nu_obj)
+        interfaces << properties_get(interface_name, nu_obj, all_intf: all_intf)
```
[the remark this answers is not in the document]
Comment (chrisvanheuveln) [reply lib/puppet/provider/cisco_interface/cisco.rb]: done

**Comment by mikewiebe** [comment]
👍

**Comment by chrisvanheuveln** [comment]
> General comment: We should probably run a full regression test with these changes just to make sure there are no hidden issues.

Full suite tested on dt-n9k5-1 and n7k-j.
- They did uncover a couple of minor bugs in NU, now fixed (https://github.com/cisco/cisco-network-node-utils/pull/636).
- There are failures with `vxlan_vtep/vxlan_vtep_vni`. I have open U.S.'s to address those.

**Pull request merged**